### ISSUE TITLE
TextField with formatters needs multiple delete taps to delete text

### ISSUE THREAD

**Issue by acoutts** [flutter/flutter#67963]
## Steps to Reproduce

1. Add a `TextInputFormatter` to a `TextField` your app.
2. Input some valid characters to the field.
3. Then input some invalid characters (which the formatter ignores).
3. Hit backspace and you'll see the input doesn't change.
4. Hit backspace repeatedly until you remove all of the phantom "bad characters, and then the input in the field will start to change again.

Unfocusing and re-focusing the TextField fixes it so the phantom characters don't exist anymore.

**Expected results:** 
When TextInputFormatter ignores some new input, the next value of newValue to `formatEditUpdate` should not use the phantom characters ignored in the last update call.

**Actual results:** 
Phantom characters are added which require additional backspaces that don't appear to remove anything.

Here's some output from `formatEditUpdate` where I print the value of `oldValue` and `newValue` on each update. Notice at some point `oldLen` stops incrementing - visually on the app, this is the point where the input formatter is now returning the same value as before (new characters are now invalid input). Despite this you see the length of `newValue` continue to increment from the last invalid length. After adding these invalid characters, it now takes 9 backspace presses until the input is actually changed, due to the phantom characters added to the field.
```
flutter: oldLen: 0 | newLen: 1
flutter: maskedValue: 4
flutter: oldLen: 1 | newLen: 2
flutter: maskedValue: 44
flutter: oldLen: 2 | newLen: 3
flutter: maskedValue: 444
flutter: oldLen: 3 | newLen: 4
flutter: maskedValue: 444-4
flutter: oldLen: 5 | newLen: 6
flutter: maskedValue: 444-44
flutter: oldLen: 6 | newLen: 7
flutter: maskedValue: 444-444
flutter: oldLen: 7 | newLen: 8
flutter: maskedValue: 444-444-4
flutter: oldLen: 9 | newLen: 10
flutter: maskedValue: 444-444-44
flutter: oldLen: 10 | newLen: 11
flutter: maskedValue: 444-444-444
flutter: oldLen: 11 | newLen: 12
flutter: maskedValue: 444-444-4444
flutter: oldLen: 12 | newLen: 13
flutter: maskedValue: 444-444-4444
flutter: oldLen: 12 | newLen: 14
flutter: maskedValue: 444-444-4444
flutter: oldLen: 12 | newLen: 15
flutter: maskedValue: 444-444-4444
flutter: oldLen: 12 | newLen: 16
flutter: maskedValue: 444-444-4444
flutter: oldLen: 12 | newLen: 17
flutter: maskedValue: 444-444-4444
flutter: oldLen: 12 | newLen: 18
flutter: maskedValue: 444-444-4444
flutter: oldLen: 12 | newLen: 19
flutter: maskedValue: 444-444-4444
flutter: oldLen: 12 | newLen: 20
flutter: maskedValue: 444-444-4444
flutter: oldLen: 12 | newLen: 20
```

Here is a minimal project reproducing the issue. It uses a `FilteringTextInputFormatter.digitsOnly`. Input some digits, then input `n` number of non-digit characters and try to hit backspace. You'll then see the phantom character issue when you try to delete those invalid characters you inputted.

https://github.com/acoutts/text_formatter_bug

<details>
  <summary>Logs</summary>

```
```

```
```

```
[✓] Flutter (Channel beta, 1.22.1, on Mac OS X 10.15.6 19G2021, locale en-US)
    • Flutter version 1.22.1 at /Users/andrewcoutts/Projects/flutter
    • Framework revision f30b7f4db9 (4 days ago), 2020-10-08 10:06:30 -0700
    • Engine revision 75bef9f6c8
    • Dart version 2.10.1

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/andrewcoutts/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/andrewcoutts/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.48.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.1

[✓] Connected device (6 available)
    • sdk gphone x86 arm (mobile)         • emulator-5554                        • android-x86    • Android 11 (API 30) (emulator)
    • xff0c-se (mobile)                   • 00008030-0016283036BB802E            • ios            • iOS 14.0.1
    • iPhone 11 Pro Max (mobile)          • 5E5512C0-5367-4BDE-8C6C-467B98B3454E • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-0 (simulator)
    • iPhone SE (2nd generation) (mobile) • F703B11E-9045-499D-87E6-A3FBF6812717 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-0 (simulator)
    • Web Server (web)                    • web-server                           • web-javascript • Flutter Tools
    • Chrome (web)                        • chrome                               • web-javascript • Google Chrome 86.0.4240.75
    ! Error: xff0c-se is busy: Making xff0c-se ready for development. Xcode will continue when xff0c-se is finished. (code -10)

• No issues found!
```

</details>

**Comment by acoutts** [comment]
Just to clarify, this affects anything which extends `TextInputFormatter`. The example app uses `FilteringTextInputFormatter.digitsOnly` to demonstrate the issue.

**Comment by pedromassangocode** [comment]
Similar to https://github.com/flutter/flutter/issues/67898 and is a duplicate of https://github.com/flutter/flutter/issues/67236 and https://github.com/flutter/flutter/issues/67828 both are closed and fixed by https://github.com/flutter/flutter/pull/67892 however the current issue is still reproducible on latest `master` channel (see image).

This was eventually introduced by https://github.com/flutter/flutter/pull/63754. It has nothing to do with the input formatters but with the text field itself (see the issues mentioned above).

This may be related to https://github.com/flutter/flutter/issues/67898 as well.

<img width="230" height="500" src="https://user-images.githubusercontent.com/67424416/95838646-ef0f8200-0d39-11eb-8141-535e0cd2dad4.gif">


<details><summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 1.23.0-14.0.pre.151, on Mac OS X 10.15.7 19H2 x86_64, locale en)
    • Flutter version 1.23.0-14.0.pre.151 at /Users/pedromassango/dev/SDKs/flutter_master
    • Framework revision 625ffcaaa6 (4 hours ago), 2020-10-13 01:07:02 -0400
    • Engine revision 6634406889
    • Dart version 2.11.0 (build 2.11.0-213.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/pedromassango/Library/Android/sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Community Edition (version 2020.2.3)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin installed
    • Dart plugin version 202.7319.5

[✓] VS Code (version 1.48.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.14.1

[✓] Connected device (4 available)
    • Android SDK built for x86 (mobile) • emulator-5554 • android-x86    • Android 10 (API 29) (emulator)
    • macOS (desktop)                    • macos         • darwin-x64     • Mac OS X 10.15.7 19H2 x86_64
    • Web Server (web)                   • web-server    • web-javascript • Flutter Tools
    • Chrome (web)                       • chrome        • web-javascript • Google Chrome 86.0.4240.80

• No issues found!
Process finished with exit code 0
```
</details>

**Comment by pedromassangocode** [comment]
Updating title to better describe the issue

**Comment by acoutts** [comment]
Thanks @pedromassango!

**Comment by xu-baolin** [comment]
Fixed via #67892

**Comment by pedromassangocode** [comment]
Yes, it seems like I tested with a version that didn’t contains the fix. The fix has landed on Flutter master with version `1.23.0-14.0.pre.152`.

I think this can be closed now.
